### PR TITLE
Review Time #3

### PR DESCRIPTION
**Pull request by namchid**
Main changes:

* Caching requested index pages (only HTML imports were cached previously)
* Updated README

### CODE REVIEW

File: _server_execute_js.js [hunk 1]
```diff
@@ -7,26 +7,36 @@ module.exports = function() {
     */
     hybridVersion: function() {
       const ipcRenderer = require('electron').ipcRenderer;
+      const remote = require('electron').remote;
+      let pc = remote.getGlobal('pageCache');
```
Comment (usergenic) [inline _server_execute_js.js]: `const pc = ` since pc is never reassigned

File: _server_execute_js.js [hunk 2]
```diff
@@ -7,26 +7,36 @@ module.exports = function() {
...
+      let cachedValue = pc.get('_true_' + document.documentElement.outerHTML);
```
Comment (usergenic) [inline _server_execute_js.js]: `const cachedValue = ` since cachedValue is never reassigned

File: _server_execute_js.js [hunk 3]
```diff
@@ -7,26 +7,36 @@ module.exports = function() {
...
 
-      if (document.querySelectorAll('link[rel="import"]').length > 0) {
-        let html = document.cloneNode(true);
-        html.querySelector('body').removeAttribute('unresolved');
+      if(cachedValue == -1) {
+        if (document.querySelectorAll('link[rel="import"]').length > 0) {
+          let html = document.cloneNode(true);
```
Comment (usergenic) [inline _server_execute_js.js]: `const html = ` since html is never reassigned.

File: _server_execute_js.js [hunk 4]
```diff
@@ -7,26 +7,36 @@ module.exports = function() {
...
     hybridVersion: function() {
       const ipcRenderer = require('electron').ipcRenderer;
+      const remote = require('electron').remote;
+      let pc = remote.getGlobal('pageCache');
+      let cachedValue = pc.get('_true_' + document.documentElement.outerHTML);
 
-      if (document.querySelectorAll('link[rel="import"]').length > 0) {
-        let html = document.cloneNode(true);
-        html.querySelector('body').removeAttribute('unresolved');
+      if(cachedValue == -1) {
```
Comment (usergenic) [inline _server_execute_js.js]: `cachedValue === -1`

The `===` in JavaScript makes sure the value is exactly what you want, whereas the `==` is a little fudgey.  (For example `"-1" == -1` and `[-1] == -1`)

File: _server_execute_js.js [hunk 5]
```diff
@@ -7,26 +7,36 @@ module.exports = function() {
...
+          let html = document.cloneNode(true);
+          html.querySelector('body').removeAttribute('unresolved');
 
-        let shadowPolymerScript = document.createElement('script');
-        shadowPolymerScript.innerText =
-            'window.Polymer = { dom: "shadow"}';
-        let imports = html.querySelectorAll('link[rel="import"]');
-        html.querySelector('head').insertBefore(shadowPolymerScript,
-          imports[0]);
+          let shadowPolymerScript = document.createElement('script');
```
Comment (usergenic) [inline _server_execute_js.js]: `const shadowPolymerScript = ` ...
I realize this is just newly indented, but pointing out const anyway :)

File: _server_execute_js.js [hunk 6]
```diff
@@ -7,26 +7,36 @@ module.exports = function() {
...
+          shadowPolymerScript.innerText =
+              'window.Polymer = { dom: "shadow"}';
+          let imports = html.querySelectorAll('link[rel="import"]');
```
Comment (usergenic) [inline _server_execute_js.js]: `const imports`

File: _server_execute_js.js [hunk 7]
```diff
@@ -35,18 +45,28 @@ module.exports = function() {
     */
     shadyVersion: function() {
       const ipcRenderer = require('electron').ipcRenderer;
+      const remote = require('electron').remote;
+      let pc = remote.getGlobal('pageCache');
+      let cachedValue = pc.get('_true_' + document.documentElement.outerHTML);
```
Comment (usergenic) [inline _server_execute_js.js]: `const pc =`
`const cachedValue = `

File: _server_execute_js.js [hunk 8]
```diff
@@ -35,18 +45,28 @@ module.exports = function() {
...
 
-      if (document.querySelectorAll('link[rel="import"]').length > 0) {
-        let html = document.cloneNode(true);
-        html.querySelector('body').removeAttribute('unresolved');
-        removeImportsAndSetInCache(html);
-        makePathsRelative(html);
+      if(cachedValue == -1) {
```
Comment (usergenic) [inline _server_execute_js.js]: `cachedValue === -1`

File: _server_execute_js.js [hunk 9]
```diff
@@ -35,18 +45,28 @@ module.exports = function() {
...
+        if (document.querySelectorAll('link[rel="import"]').length > 0) {
+          let html = document.cloneNode(true);
```
Comment (usergenic) [inline _server_execute_js.js]: `const html =`

File: _server_execute_js.js [hunk 10]
```diff
@@ -250,31 +270,41 @@ module.exports = function() {
       }
 
       const ipcRenderer = require('electron').ipcRenderer;
+      const remote = require('electron').remote;
+      let pc = remote.getGlobal('pageCache');
+      let cachedValue = pc.get('_false_' + document.documentElement.outerHTML);
```
Comment (usergenic) [inline _server_execute_js.js]: `const pc =`
`const cachedValue =`

File: _server_execute_js.js [hunk 11]
```diff
@@ -250,31 +270,41 @@ module.exports = function() {
...
-      replaceShadowRoots(doc, clonedDoc);
-      removeImports(clonedDoc);
-      hashKey = hash(hashKey);
-      if(hashValue != '') {
-        ipcRenderer.send('setAsyncImports', hashKey, hashValue);
-      }
-      removeScripts(clonedDoc);
-      insertShadowStyles(clonedDoc, shadowStyleList);
-      insertScriptsAndImports(clonedDoc, hashKey);
+      if(cachedValue == -1) {
```
Comment (usergenic) [inline _server_execute_js.js]: `cachedValue === -1`

**Review by usergenic — CHANGES_REQUESTED** [review]
My change requests here are fairly minor.

The `const`-ing stuff is more about correctness and intention-revealing for the vars.

The `===` business will burn you eventually so its definitely worth fixing.

File: _server_execute_js.js (hunk 5 above)
Comment (namchid) [inline _server_execute_js.js]: Haha, yes, thanks for checking even the old stuff!

File: _server_execute_js.js (hunk 1 above)
Comment (namchid) [inline _server_execute_js.js]: Addressed the comments in commits 6158051 and 69744b1. I also went back and checked/modified the other instances of `let`s and `==`s.

**Pull request merged**